apps/simplex-chat/Simplex/Chat.hs

```diff
@@ -61,66 +72,136 @@ inputSubscriber = do
               SendMessage c msg -> showSentMessage c msg
               _ -> printToView [plain s]
             runExceptT (processChatCommand cmd) >>= \case
-              Left (ChatErrorAgent c e) -> showAgentError c e
+              Left e -> showChatError e
               _ -> pure ()
 
 processChatCommand :: ChatMonad m => ChatCommand -> m ()
 processChatCommand = \case
   ChatHelp -> printToView chatHelpInfo
   MarkdownHelp -> printToView markdownInfo
-  AddContact c -> do
-    (_, qInfo) <- withAgent c (`createConnection` Just (fromContact c))
-    showInvitation c qInfo
-  Connect c qInfo ->
-    void . withAgent c $ \smp -> joinConnection smp (Just $ fromContact c) qInfo
-  DeleteContact c -> do
-    withAgent c (`deleteConnection` fromContact c)
-    showContactDeleted c
-    unsetActive' $ ActiveC c
-  SendMessage c msg -> do
-    void . withAgent c $ \smp -> sendMessage smp (fromContact c) msg
-    setActive' $ ActiveC c
+  AddContact cRef -> do
+    (connId, qInfo) <- withAgent (fromMaybe "" cRef) createConnection
+    userId <- asks currentUserId
+    contact <- withStore $ \st -> createDirectContact st userId connId cRef
+    showInvitation (localContactRef contact) qInfo
+  Connect cRef qInfo -> do
+    userId <- asks currentUserId
+    connId <- withAgent (fromMaybe "" cRef) $ \agent -> joinConnection agent qInfo "user profile here"
+    void $ withStore $ \st -> createDirectContact st userId connId cRef
+  DeleteContact cRef -> do
+    userId <- asks currentUserId
+    conns <- withStore $ \st -> getContactConnections st userId cRef
+    withAgent cRef $ \smp -> forM_ conns $ \Connection {agentConnId} ->
+      deleteConnection smp agentConnId `catchError` \(_ :: AgentErrorType) -> pure ()
+    void $ withStore $ \st -> deleteContact st userId cRef
+    unsetActive $ ActiveC cRef
+    when (null conns) . throwError . ChatErrorContact $ CENotFound cRef
+    showContactDeleted cRef
+  SendMessage cRef msg -> do
+    userId <- asks currentUserId
+    Connection {agentConnId} <- withStore $ \st -> getContactConnection st userId cRef
+    let body = MsgBodyContent {contentType = SimplexContentType XCText, contentHash = Nothing, contentData = MBFull $ MsgData msg}
+        rawMsg = rawChatMessage ChatMessage {chatMsgId = Nothing, chatMsgEvent = XMsgNew MTText, chatMsgBody = [body], chatDAGIdx = Nothing}
+    void . withAgent cRef $ \smp -> sendMessage smp agentConnId $ serializeRawChatMessage rawMsg
+    setActive $ ActiveC cRef
```

Comment:
is it correct that notion of active contact is needed only for terminal chat (and not other UI)? if so shouldn't it be decoupled from chat logic, e.g. by a way of terminal component changing it?

Reply:
I would worry about it later